eth/protocols/snap/sync.go

```diff
@@ -2899,3 +2900,76 @@ func (s *capacitySort) Swap(i, j int) {
 	s.ids[i], s.ids[j] = s.ids[j], s.ids[i]
 	s.caps[i], s.caps[j] = s.caps[j], s.caps[i]
 }
+
+// healRequestSort implements the Sort interface, allowing sorting trienode
+// heal requests, which is a prerequisite for merging storage-requests.
+type healRequestSort struct {
+	hashes []common.Hash
+	paths  []trie.SyncPath
+}
+
+func (t *healRequestSort) Len() int {
+	return len(t.hashes)
+}
+
+func (t *healRequestSort) Less(i, j int) bool {
+	a := t.paths[i]
+	b := t.paths[j]
+	switch bytes.Compare(a[0], b[0]) {
+	case -1:
+		return true
+	case 1:
+		return false
+	}
+	// identical first part
+	if len(a) < len(b) {
+		return true
+	}
+	if len(b) < len(a) {
+		return false
+	}
+	if len(a) == 2 {
+		return bytes.Compare(a[1], b[1]) < 0
+	}
+	return false
+}
+
+func (t *healRequestSort) Swap(i, j int) {
+	t.hashes[i], t.hashes[j] = t.hashes[j], t.hashes[i]
+	t.paths[i], t.paths[j] = t.paths[j], t.paths[i]
+}
+
+// Merge merges the pathsets, so that several storage requests concerning the
+// same account are merged into one, to reduce bandwidth.
+// OBS: This operation is moot if t has not first been sorted.
+func (t *healRequestSort) Merge() []TrieNodePathSet {
+	var nPathset []TrieNodePathSet
+	var last TrieNodePathSet
+
+	for _, path := range t.paths {
+		pSet := TrieNodePathSet([][]byte(path))
+		if len(path) == 1 {
+			nPathset = append(nPathset, pSet)
+			last = nil
+			continue
+		}
+		if last != nil && bytes.Equal(pSet[0], last[0]) {
+			// we can merge this into the last one
+			last = append(last, pSet[1])
+			nPathset[len(nPathset)-1] = last
+		} else {
+			nPathset = append(nPathset, pSet)
+			last = pSet
+		}
```

Comment:
Small logical nitpick. The code does the append clause first and the init branch second. Whilst the code is correct, it's harder to understand because you need to jump ahead to read the else branch before you can intepret the condition and the if clause. I've suggest reversing the two:

```go
if last == nil || !bytes.Equal(pSet[0], last[0]) {
	nPathset = append(nPathset, pSet)
	last = pSet
} else {
	// We can merge this into the last one
	last = append(last, pSet[1])
	nPathset[len(nPathset)-1] = last
}
```

But if you feel it's only me being picky, the original code is fine too.

Reply:
I'm with you, but would go a step farther and un-indent the last clause